ambry-router/src/main/java/com/github/ambry/router/GetBlobOperation.java

```diff
@@ -140,18 +144,93 @@ class GetBlobOperation extends GetOperation {
       Callback<GetBlobResultInternal> callback, RouterCallback routerCallback, BlobIdFactory blobIdFactory,
       KeyManagementService kms, CryptoService cryptoService, CryptoJobHandler cryptoJobHandler, Time time,
-      boolean isEncrypted, QuotaChargeCallback quotaChargeCallback) {
+      boolean isEncrypted, QuotaChargeCallback quotaChargeCallback, AmbryCache blobMetadataCache) {
     super(routerConfig, routerMetrics, clusterMap, responseHandler, blobId, options, callback, kms, cryptoService,
         cryptoJobHandler, time, isEncrypted);
     this.routerCallback = routerCallback;
     this.blobIdFactory = blobIdFactory;
     this.quotaChargeCallback = quotaChargeCallback;
-    firstChunk = new FirstGetChunk();
+    this.blobMetadataCache = blobMetadataCache;
+    BlobMetadata blobMetadata = null;
+    if (shouldLookupMetadataCache()) {
+      blobMetadata = (BlobMetadata) blobMetadataCache.getObject(blobId.toString());
+    }
+    if (blobMetadata != null) {
+      firstChunk = new CachedFirstChunk(blobMetadata);
+    } else {
+      firstChunk = new FirstGetChunk();
+    }
+  }
+
+  /**
+   * Decides if we should look up the metadata cache during a GET request
+   * @return True if and only if
+   *  - cache exists
+   *  - it is a RANGE request
+   *  - it is not a raw mode request
+   *  - the blobID is for a metadata blob
+   *  - the cache not yet been consulted
+   *  False, otherwise.
+   */
+  boolean shouldLookupMetadataCache() {
+    return (blobMetadataCache != null &&
+        options.getBlobOptions.getRange() != null &&
+        options.getBlobOptions.isRawMode() == false &&
+        blobId.getBlobDataType() == BlobId.BlobDataType.METADATA);
+  }
+
+  /**
+   * Decides if we should cache metadata during a GET request
+   * This method should be called after decryption is complete.
+   * @return True if and only if
+   *  - cache exists
+   *  - it is RANGE request
+   *  - it is not a raw mode request
+   *  - the blobID is for a metadata blob
+   *  - the cache was looked up and we got a cache miss
+   *  False, otherwise.
+   */
+  boolean shouldSaveMetadata() {
+    return (blobMetadataCache != null &&
+        options.getBlobOptions.getRange() != null &&
+        options.getBlobOptions.isRawMode() == false &&
+        blobId.getBlobDataType() == BlobId.BlobDataType.METADATA &&
+        totalSize >= routerConfig.routerSmallestBlobForMetadataCache);
+  }
+
+  /**
+   * Conditionally saves blob metadata for composite blobs
+   * @return True if metadata was saved successfully, else False.
+   */
+  boolean saveMetadata() {
+    boolean put_result = false;
+    if (shouldSaveMetadata()) {
+      BlobMetadata blobMetadata = new BlobMetadata(blobId.toString(), blobInfo, compositeBlobInfo);
+      put_result = blobMetadataCache.putObject(blobMetadata.getBlobId(), blobMetadata);
+      logger.info(String.format("Saving metadata for blobId = %s, result = %s", blobId, put_result));
```

Comment:
do logger.trace please. And you don't have to call String.format, logger's methods by default do string formatting for you.